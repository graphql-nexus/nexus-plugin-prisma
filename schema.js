module.exports = require('./dist/schema-entrypoint')
