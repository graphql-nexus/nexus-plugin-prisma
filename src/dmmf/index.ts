export { get } from './importer'
export { transform } from './transformer'
export { DMMFClass as DMMF } from './DMMFClass'
export { ExternalDMMF as External } from './types'
