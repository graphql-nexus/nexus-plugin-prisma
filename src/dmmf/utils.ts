import * as Photon from '@prisma/photon'
import { transform } from './transformer'
import { DMMFClass } from './DMMFClass'

export function fromPhotonDMMF(photonDMMF: Photon.DMMF.Document): DMMFClass {
  return new DMMFClass(transform(photonDMMF))
}

export const get = (photonClientPackagePath: string): DMMFClass => {
  let photonClientPackage
  try {
    photonClientPackage = require(photonClientPackagePath)
  } catch (error) {
    throw new Error(
      `Could not find photon package at ${photonClientPackagePath}. Check that you have configured your Photon generator block in schema.prisma correctly and run prisma generate.`,
    )
  }
  return fromPhotonDMMF(photonClientPackage.dmmf)
}
