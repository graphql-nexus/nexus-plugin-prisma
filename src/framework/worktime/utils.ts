import * as Prisma from '@prisma/sdk'
import chalk from 'chalk'
import { WorktimeLens } from 'nexus/plugin'
import stripAnsi from 'strip-ansi'
import * as Scaffolders from './scaffolders'

/**
 * Pinned query-engine version. Calculated at build time and based on `@prisma/cli` version
 */
export const PRISMA_QUERY_ENGINE_VERSION: string = require('@prisma/cli/package.json').prisma.version

/**
 * Get the declared generator blocks in the user's PSL file.
 *
 * If Prisma SDK throws a no-models-found error then it is caught, a model is
 * scaffolded, and this function is re-run.
 */
export async function getGenerators(nexus: WorktimeLens, schemaPath: string): Promise<Prisma.Generator[]> {
  return Prisma.getGenerators({
    schemaPath,
    printDownloadProgress: false,
    version: PRISMA_QUERY_ENGINE_VERSION,
  })
    .catch(async (e: Error) => {
      if (isNoModelsDefinedError(e)) {
        await Scaffolders.exampleModelBlock(schemaPath)
        nexus.log.warn(
          `An example model has been scaffolded for you in ${chalk.bold(
            nexus.layout.projectRelative(schemaPath)
          )}`
        )
        return getGenerators(nexus, schemaPath)
      }
      return Promise.reject(e)
    })
    .then(async (generators) => {
      const hasClientGenerator = generators.find((g) => g.options?.generator.provider === 'prisma-client-js')
      if (!hasClientGenerator) {
        await Scaffolders.prismaClientGeneratorBlock(schemaPath)
        nexus.log.warn(
          `A Prisma Client generator block has been scaffolded for you in ${chalk.bold(
            nexus.layout.projectRelative(schemaPath)
          )}`
        )
        // TODO: Generate it programmatically instead for performance reason
        return getGenerators(nexus, schemaPath)
      } else {
        return generators
      }
    })
}

/**
 * Check if the thrown error is of NoModelsDefined type.
 */
function isNoModelsDefinedError(e: Error): boolean {
  return Boolean(stripAnsi(e.message).match(/.*You don't have any models defined.*/))
}
