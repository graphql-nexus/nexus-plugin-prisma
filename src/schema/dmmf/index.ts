export * from './utils'
export * from './transformer'
export * from './DmmfDocument'
export * from './DmmfTypes'
