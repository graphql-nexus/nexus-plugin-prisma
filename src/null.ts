import { DmmfDocument, DmmfTypes } from './dmmf'

/**
 * Take the incoming GraphQL args of a resolver and replaces all `null` values
 * that maps to a non-nullable field in the Prisma Schema, by `undefined` values.
 * 
 * In Prisma, a `null` value has a specific meaning for the underlying database.
 * Therefore, `undefined` is used instead to express the optionality of a field.
 * 
 * In GraphQL however, no difference is made between `null` and `undefined`.
 * This is the reason why we need to convert all `null` values that were assigned to `non-nullable` fields to `undefined`.
 */
export function transformNullsToUndefined(
  args: Record<string, any>,
  schemaArgs: Record<string, DmmfTypes.SchemaArg>,
  dmmf: DmmfDocument,
) {
  const keys = Object.keys(args)
  for (const key of keys) {
    const val = args[key]
    const schemaArg = schemaArgs[key]

    if (!schemaArg) {
      throw new Error(`Could not find schema arg with name: ${key}`)
    }

    const shouldConvertNullToUndefined =
      val === null && schemaArg.inputType.isNullable === false

    if (shouldConvertNullToUndefined) {
      args[key] = undefined
    } else if (isObject(val)) {
      const newSchemaArgs = dmmf.getInputTypeWithIndexedFields(
        schemaArg.inputType.type,
      ).fields

      args[key] = transformNullsToUndefined(args[key], newSchemaArgs, dmmf)
    }
  }

  return args
}

function isObject(obj: any): boolean {
  return obj && typeof obj === 'object'
}
