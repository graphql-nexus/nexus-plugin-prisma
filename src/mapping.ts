import * as DMMF from './dmmf'
import {
  defaultFieldNamingStrategy,
  FieldNamingStrategy,
  OperationName,
} from './naming-strategies'
import { flatMap } from './utils'

interface BaseMappedField {
  field: string
  operation: OperationName
  model: string
  photonAccessor: string
}

export interface MappedField extends Omit<BaseMappedField, 'field'> {
  field: DMMF.Data.SchemaField
}

const buildField = (
  mapping: DMMF.Data.Mapping,
  operation: OperationName,
): BaseMappedField => ({
  operation,
  field: mapping[operation],
  model: mapping.model,
  photonAccessor: mapping.plural,
})

const CRUD_MAPPED_FIELDS: Record<
  string,
  (m: DMMF.Data.Mapping) => BaseMappedField[]
> = {
  Query: m => [buildField(m, 'findOne'), buildField(m, 'findMany')],
  Mutation: m => [
    buildField(m, 'create'),
    buildField(m, 'update'),
    buildField(m, 'updateMany'),
    buildField(m, 'delete'),
    buildField(m, 'deleteMany'),
    buildField(m, 'upsert'),
  ],
}

export const getCrudMappedFields = (
  typeName: 'Query' | 'Mutation',
  dmmf: DMMF.DMMF,
  namingStrategy: FieldNamingStrategy = defaultFieldNamingStrategy,
): MappedField[] =>
  flatMap(dmmf.mappings, m => CRUD_MAPPED_FIELDS[typeName](m)).map(
    mappedField => ({
      ...mappedField,
      field: {
        ...dmmf.getOutputType(typeName).getField(mappedField.field),
        name: namingStrategy[mappedField.operation](
          mappedField.field,
          mappedField.model,
        ),
      },
    }),
  )
