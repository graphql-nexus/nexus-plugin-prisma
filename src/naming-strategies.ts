import camelCase from 'camelcase'
import pluralize from 'pluralize'
import { InternalDMMF } from './dmmf'
import { upperFirst } from './utils'

export interface ArgsNamingStrategy {
  whereInput: (typeName: string, fieldName: string) => string
  orderByInput: (typeName: string, fieldName: string) => string
  relationFilterInput: (typeName: string, fieldName: string) => string
}

export const defaultArgsNamingStrategy: ArgsNamingStrategy = {
  whereInput(typeName, fieldName) {
    return `${upperFirst(typeName)}${upperFirst(fieldName)}WhereInput`
  },
  orderByInput(typeName, fieldName) {
    return `${upperFirst(typeName)}${upperFirst(fieldName)}OrderByInput`
  },
  relationFilterInput(typeName, fieldName) {
    return `${upperFirst(typeName)}${upperFirst(fieldName)}Filter`
  },
}

export type OperationName = Exclude<keyof InternalDMMF.Mapping, 'model' | 'plural'>

export type FieldNamingStrategy = Record<OperationName, (fieldName: string, modelName: string) => string>

export const defaultFieldNamingStrategy: FieldNamingStrategy = {
  // findFirst: (_, modelName) => camelCase(modelName),
  aggregate: (fieldName) => fieldName,
  findUnique: (_, modelName) => camelCase(modelName),
  findMany: (_, modelName) => camelCase(pluralize(modelName)),
  groupBy: (fieldName) => fieldName,
  create: (fieldName) => fieldName,
  createMany: (fieldName) => fieldName,
  delete: (fieldName) => fieldName,
  deleteMany: (fieldName) => fieldName,
  update: (fieldName) => fieldName,
  updateMany: (fieldName) => fieldName,
  upsert: (fieldName) => fieldName,
}
