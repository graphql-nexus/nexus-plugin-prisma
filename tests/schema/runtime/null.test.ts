import { DmmfTypes, DmmfDocument } from '../../../src/schema/dmmf'
import { getCrudMappedFields } from '../../../src/schema/mapping'
import { OperationName } from '../../../src/schema/naming-strategies'
import { transformNullsToUndefined } from '../../../src/schema/null'
import { indexBy } from '../../../src/schema/utils'
import { getDmmf } from '../__utils'

const operationToRoot: Record<OperationName, 'Query' | 'Mutation'> = {
  findOne: 'Query',
  findMany: 'Query',
  create: 'Mutation',
  delete: 'Mutation',
  deleteMany: 'Mutation',
  update: 'Mutation',
  updateMany: 'Mutation',
  upsert: 'Mutation',
}

async function getSchemaArgsForCrud(
  datamodel: string,
  model: string,
  operation: OperationName
): Promise<{
  schemaArgs: Record<string, DmmfTypes.SchemaArg>
  dmmf: DmmfDocument
}> {
  const dmmf = await getDmmf(datamodel)
  const mappedField = getCrudMappedFields(operationToRoot[operation], dmmf).find(
    (x) => x.operation === operation && x.model === model
  )

  if (!mappedField) {
    throw new Error(`Could not find mapped fields for model ${model} and operation ${operation}`)
  }

  return {
    schemaArgs: indexBy(mappedField.field.args, 'name'),
    dmmf,
  }
}

test('findMany: converts nulls to undefined when fields are not nullable', async () => {
  const datamodel = `
datasource db {
  provider = "postgresql"
  url      = "postgresql://"
}

model User {
  id        String   @default(cuid()) @id
  email     String?  @unique
  birthDate DateTime
  posts     Post[]
}

model Post {
  id      String @default(cuid()) @id
  authors User[] @relation(references: [id])
}
  `
  const { dmmf, schemaArgs } = await getSchemaArgsForCrud(datamodel, 'User', 'findMany')
  const incomingArgs = {
    before: null, // not nullable
    after: null, // not nullable
    first: 1,
    orderBy: {
      email: null, // nullable
      birthDate: null, // nullable
      id: 'asc',
    },
    where: {
      AND: null, // not nullable
      NOT: {
        AND: {
          birthDate: null, // not nullable
        },
        posts: null,
      },
    },
  }

  const result = transformNullsToUndefined(incomingArgs, schemaArgs, dmmf)

  expect(result).toMatchSnapshot()
})

test('create: converts nulls to undefined when fields are not nullable', async () => {
  const datamodel = `
datasource db {
  provider = "postgresql"
  url      = "postgresql://"
}

model User {
  id        String   @default(cuid()) @id
  email     String?  @unique
  birthDate DateTime
  posts     Post[]
}

model Post {
  id      String @default(cuid()) @id
  authors User[] @relation(references: [id])
}
  `
  const { dmmf, schemaArgs } = await getSchemaArgsForCrud(datamodel, 'User', 'create')
  const incomingArgs = {
    data: {
      id: null, // not nullable
      posts: {
        connect: null, // not nullable
        create: {
          id: 'titi',
        },
      },
    },
  }

  const result = transformNullsToUndefined(incomingArgs, schemaArgs, dmmf)

  expect(result).toMatchSnapshot()
})

test('model filtering: converts nulls to undefined when fields are not nullable', async () => {
  const datamodel = `
datasource db {
  provider = "postgresql"
  url      = "postgresql://"
}

model User {
  id        String   @default(cuid()) @id
  email     String?  @unique
  birthDate DateTime
  posts     Post[]
}

model Post {
  id      String @default(cuid()) @id
  authors User[] @relation(references: [id])
}
  `
  const dmmf = await getDmmf(datamodel)
  const schemaArgs = dmmf.getOutputType('User').fields.find((f) => f.name === 'posts')?.args!
  const indexedSchemaArgs = indexBy(schemaArgs, 'name')
  const incomingArgs = {
    where: {
      id: null, // not nullable
      authors: {
        every: {
          email: null,
          birthDate: null, // not nullable
          posts: null,
        },
      },
    },
  }

  const result = transformNullsToUndefined(incomingArgs, indexedSchemaArgs, dmmf)

  expect(result).toMatchSnapshot()
})

test('do not convert args that are arrays', async () => {
  const datamodel = `
datasource db {
  provider = "postgresql"
  url      = "postgresql://"
}

model User {
  id        String   @default(cuid()) @id
  email     String?  @unique
  birthDate DateTime
  posts     Post[]
}

model Post {
  id      String @default(cuid()) @id
  authors User[] @relation(references: [id])
}
  `
  const dmmf = await getDmmf(datamodel)
  const schemaArgs = dmmf.getOutputType('User').fields.find((f) => f.name === 'posts')?.args!
  const indexedSchemaArgs = indexBy(schemaArgs, 'name')
  const incomingArgs = {
    where: {
      OR: [{ something: true }, { something: false }]
    },
  }

  const result = transformNullsToUndefined(incomingArgs, indexedSchemaArgs, dmmf)

  expect(result).toMatchSnapshot()
})
