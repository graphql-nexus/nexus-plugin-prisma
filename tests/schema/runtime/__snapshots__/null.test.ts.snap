// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create: converts nulls to undefined when fields are not nullable 1`] = `
Object {
  "data": Object {
    "id": undefined,
    "posts": Object {
      "connect": undefined,
      "create": Object {
        "id": "titi",
      },
    },
  },
}
`;

exports[`do not convert args that are arrays 1`] = `
Object {
  "where": Object {
    "OR": Array [
      Object {
        "something": true,
      },
      Object {
        "something": false,
      },
    ],
  },
}
`;

exports[`findMany: converts nulls to undefined when fields are not nullable 1`] = `
Object {
  "after": undefined,
  "before": undefined,
  "first": 1,
  "orderBy": Object {
    "birthDate": undefined,
    "email": undefined,
    "id": "asc",
  },
  "where": Object {
    "AND": undefined,
    "NOT": Object {
      "AND": Object {
        "birthDate": undefined,
      },
      "posts": undefined,
    },
  },
}
`;

exports[`model filtering: converts nulls to undefined when fields are not nullable 1`] = `
Object {
  "where": Object {
    "authors": Object {
      "every": Object {
        "birthDate": undefined,
        "email": undefined,
        "posts": undefined,
      },
    },
    "id": undefined,
  },
}
`;
